tmt/steps/__init__.py

```diff
@@ -538,76 +684,174 @@ def methods(cls) -> List[Method]:
     def delegate(
             cls,
             step: Step,
-            data: StepData) -> 'BasePlugin':
+            data: Optional[StepData] = None,
+            raw_data: Optional[_RawStepData] = None) -> 'BasePlugin':
         """
         Return plugin instance implementing the data['how'] method
 
         Supports searching by method prefix as well (e.g. 'virtual').
         The first matching method with the lowest 'order' wins.
         """
+
+        if data is not None:
+            how = data.how
+        elif raw_data is not None:
+            how = raw_data['how']
+        else:
+            raise tmt.utils.GeneralError('Either data or raw data must be given')
+
+        step.debug(
+            f'{cls.__name__}.delegate(step={step}, data={data}, raw_data={raw_data})',
+            level=3)
+
         # Filter matching methods, pick the one with the lowest order
         for method in cls.methods():
             assert method.class_ is not None
-            if method.name.startswith(data['how']):
+            if method.name.startswith(how):
                 step.debug(
                     f"Using the '{method.class_.__name__}' plugin "
-                    f"for the '{data['how']}' method.", level=2)
-                plugin = method.class_(step, data)
+                    f"for the '{how}' method.", level=2)
+
+                plugin_class = method.class_
+                plugin_data_class = plugin_class._data_class
+
+                # If we're given raw data, construct a step data instance, applying
+                # normalization in the process.
+                if raw_data is not None:
+                    try:
+                        data = plugin_data_class.from_raw(raw_data, step)
+
+                    except Exception as exc:
+                        # import traceback
+                        # traceback.print_exc()
```

Comment:
Debugging leftover?

Reply:
Exactly, will fix.

Reply:
Fixed in [8d096fd](https://github.com/teemtee/tmt/pull/1461/commits/8d096fd6dedfb1f4dab9cea49f75946075af5836)